File: backend/app/utils/pagination.py
import math
from typing import Tuple


def compute_total_pages(total_items: int, page_size: int) -> int:
    safe_total = max(0, int(total_items))
    safe_page_size = max(1, int(page_size))
    return max(1, math.ceil(safe_total / safe_page_size))


def clamp_page(page: int, total_pages: int) -> int:
    if total_pages <= 1:
        return 1
    return max(1, min(int(page), int(total_pages)))


def normalize_pagination(total_items: int, page: int, page_size: int) -> Tuple[int, int, int]:
    total_pages = compute_total_pages(total_items=total_items, page_size=page_size)
    safe_page = clamp_page(page=page, total_pages=total_pages)
    offset = (safe_page - 1) * max(1, int(page_size))
    return safe_page, total_pages, offset

File: frontend-admin/src/utils/pagination.ts
export const computeTotalPages = (totalItems: number, pageSize: number): number => {
    const safeTotal = Number.isFinite(totalItems) ? Math.max(0, Math.floor(totalItems)) : 0;
    const safePageSize = Number.isFinite(pageSize) ? Math.max(1, Math.floor(pageSize)) : 1;
    return Math.max(1, Math.ceil(safeTotal / safePageSize));
};

export const clampPage = (page: number, totalPages: number): number => {
    const safeTotalPages = Number.isFinite(totalPages) ? Math.max(1, Math.floor(totalPages)) : 1;
    const safePage = Number.isFinite(page) ? Math.floor(page) : 1;
    return Math.min(Math.max(safePage, 1), safeTotalPages);
};

export const validatePageSize = (
    value: string | number,
    maxAllowed: number = 9999
): { valid: true; pageSize: number } | { valid: false; error: string } => {
    const raw = String(value).trim();
    if (!/^\d+$/.test(raw)) {
        return { valid: false, error: 'Page size must be an integer.' };
    }

    const pageSize = Number(raw);
    if (!Number.isInteger(pageSize)) {
        return { valid: false, error: 'Page size must be an integer.' };
    }
    if (pageSize < 1) {
        return { valid: false, error: 'Page size must be at least 1.' };
    }
    if (pageSize > maxAllowed) {
        return { valid: false, error: `Page size cannot exceed ${maxAllowed}.` };
    }

    return { valid: true, pageSize };
};
